framework/test/unit/engine/generator/generator.spec.ts

```diff
@@ -808,5 +809,49 @@ describe('generator', () => {
 				expect(generator['_consensus'].certifySingleCommit).not.toHaveBeenCalled();
 			});
 		});
+
+		describe('when previous finalized height change and maxRemovalHeight is non zero', () => {
+			beforeEach(async () => {
+				await generator.init({
+					blockchainDB,
+					generatorDB,
+					logger,
+					genesisHeight: 0,
+				});
+				await generator.start();
+				jest.spyOn(generator, '_handleFinalizedHeightChanged' as never);
+				jest.spyOn(generator, '_certifySingleCommitForChangedHeight' as never);
+			});
+
+			afterEach(async () => {
+				await generator.stop();
+			});
+
+			it('should not call certifySingleCommit when getMaxRemovalHeight is higher than next finalized height', async () => {
+				jest.spyOn(consensus, 'getMaxRemovalHeight').mockResolvedValue(30000);
+				generator['_consensus'].events.emit(CONSENSUS_EVENT_FINALIZED_HEIGHT_CHANGED, {
+					from: 0,
+					to: 25520,
+				});
+				await new Promise(process.nextTick);
+
+				expect(generator['_handleFinalizedHeightChanged']).toHaveBeenCalledWith(30000, 25520);
+				expect(generator['_certifySingleCommitForChangedHeight']).not.toHaveBeenCalled();
+			});
+
+			it('should call certifySingleCommit when getMaxRemovalHeight is lower than next finalized height', async () => {
+				jest.spyOn(consensus, 'getMaxRemovalHeight').mockResolvedValue(30000);
+				generator['_consensus'].events.emit(CONSENSUS_EVENT_FINALIZED_HEIGHT_CHANGED, {
+					from: 30001,
+					to: 30003,
+				});
+				await new Promise(process.nextTick);
+
+				expect(generator['_handleFinalizedHeightChanged']).toHaveBeenCalledWith(30001, 30003);
+				expect(generator['_certifySingleCommitForChangedHeight']).toHaveBeenCalledTimes(
+					1 * generator['_keypairs'].size,
+				);
+			});
+		});
```

Comment:
Perhaps a bit more cases could be added just to be sure (optional). Examples:

 - `h1 < getMaxRemovalHeight() < h2`: then commit for h2 should be created and only for heights in [getMaxRemovalHeight(), h2-1] such that `existBFTParametersExist(height+1)`.
 - `getMaxRemovalHeight() = h2-1` (special case of the previous, but I would check): only commit for height `h2`
 - `getMaxRemovalHeight() = h2`: No commits at all.

I would try all those because we have a lot of +1 or -1s and in the past we made similar mistakes that were not caught by unit tests.

Reply:
I think the other cases are tested in the original cases. Here, the main thing to test was to ensure maxRemovalHeight is used properly